clients/src/test/java/org/apache/kafka/clients/FetchSessionHandlerTest.java

```diff
@@ -0,0 +1,361 @@
+        handler.handleResponse(resp);
+
+        FetchSessionHandler.Builder builder2 = handler.newBuilder();
+        builder2.add(new TopicPartition("foo", 0),
+            new FetchRequest.PartitionData(0, 100, 200));
+        FetchSessionHandler.FetchRequestData data2 = builder2.build();
+        assertEquals(INVALID_SESSION_ID, data2.metadata().id());
+        assertEquals(INITIAL_EPOCH, data2.metadata().epoch());
+        assertMapsEqual(reqMap(new ReqEntry("foo", 0, 0, 100, 200)),
+            data.toSend(), data.sessionPartitions());
+    }
+
+    /**
+     * Test handling an incremental fetch session.
+     */
+    @Test
+    public void testIncrementals() throws Exception {
+        FetchSessionHandler handler = new FetchSessionHandler(LOG_CONTEXT, 1);
+        FetchSessionHandler.Builder builder = handler.newBuilder();
+        builder.add(new TopicPartition("foo", 0),
+            new FetchRequest.PartitionData(0, 100, 200));
+        builder.add(new TopicPartition("foo", 1),
+            new FetchRequest.PartitionData(10, 110, 210));
+        FetchSessionHandler.FetchRequestData data = builder.build();
+        assertMapsEqual(reqMap(new ReqEntry("foo", 0, 0, 100, 200),
+            new ReqEntry("foo", 1, 10, 110, 210)),
+            data.toSend(), data.sessionPartitions());
+        assertEquals(INVALID_SESSION_ID, data.metadata().id());
+        assertEquals(INITIAL_EPOCH, data.metadata().epoch());
+
+        FetchResponse resp = new FetchResponse(Errors.NONE,
+            respMap(new RespEntry("foo", 0, 10, 20),
+                    new RespEntry("foo", 1, 10, 20)),
+            0, 123);
+        handler.handleResponse(resp);
+
+        // Test an incremental fetch request which adds one partition and modifies another.
+        FetchSessionHandler.Builder builder2 = handler.newBuilder();
+        builder2.add(new TopicPartition("foo", 0),
+            new FetchRequest.PartitionData(0, 100, 200));
+        builder2.add(new TopicPartition("foo", 1),
+            new FetchRequest.PartitionData(10, 120, 210));
+        builder2.add(new TopicPartition("bar", 0),
+            new FetchRequest.PartitionData(20, 200, 200));
+        FetchSessionHandler.FetchRequestData data2 = builder2.build();
+        assertFalse(data2.metadata().isFull());
+        assertMapEquals(reqMap(new ReqEntry("foo", 0, 0, 100, 200),
+                new ReqEntry("foo", 1, 10, 120, 210),
+                new ReqEntry("bar", 0, 20, 200, 200)),
+            data2.sessionPartitions());
+        assertMapEquals(reqMap(new ReqEntry("bar", 0, 20, 200, 200),
+                new ReqEntry("foo", 1, 10, 120, 210)),
+            data2.toSend());
+
+        FetchResponse resp2 = new FetchResponse(Errors.NONE,
+            respMap(new RespEntry("foo", 1, 20, 20)),
+            0, 123);
+        handler.handleResponse(resp2);
+        assertMapsEqual(reqMap(new ReqEntry("foo", 0, 0, 100, 200)),
+            data.toSend(), data.sessionPartitions());
```

Comment:
Hmm, are we supposed to test data? Should we build a new request?

Reply:
In this case, it was intentional to skip building a new request.  I'll add a comment to make it clearer (also we don't need to test `data#toSend` again)